pkg/distributor/push.go

```diff
@@ -141,12 +147,42 @@ func handler(
 			if code != 202 {
 				level.Error(logger).Log("msg", "push error", "err", err)
 			}
-			addHeaders(w, err)
+			addHeaders(w, err, r, code, retryCfg, logger)
 			http.Error(w, msg, code)
 		}
 	})
 }
 
+/*
+calculateRetryAfter derives the value for the "Retry-After" header based on the provided HTTP request and RetryConfig.
+
+It calculates the delay in seconds for retry attempts, considering factors such as the Retry-Attempt header,
+RetryConfig parameters, and exponential backoff strategy. If Retry-Attempt is not valid or less than 1, it defaults to 1.
+If Retry-Attempt exceeds MaxAllowedAttempts from RetryConfig, it is capped at MaxAllowedAttempts.
+
+The delaySeconds is then calculated as a random number within the range of exponential backoff between minRetry and maxRetry.
+
+Returns: A string representing the calculated value for the "Retry-After" header.
+*/
+func calculateRetryAfter(req *http.Request, retryCfg *RetryConfig) string {
+	retryAttempt, err := strconv.Atoi(req.Header.Get("Retry-Attempt"))
+	// If retry-attempt is not valid, set it to default 1
+	if err != nil || retryAttempt < 1 {
+		retryAttempt = 1
+	}
+	if retryAttempt > retryCfg.MaxAllowedAttempts {
+		retryAttempt = retryCfg.MaxAllowedAttempts
+	}
+
+	var minRetry, maxRetry int64
+	minRetry = int64(retryCfg.Base.Seconds() * math.Pow(2, float64(retryAttempt-1)))
+	maxRetry = int64(retryCfg.Base.Seconds() * math.Pow(2, float64(retryAttempt)))
```

Comment:
We can replace `math.Pow(2, ...)` with integer shifts: `1 << retryAttempt`, which will be much faster.

Reply:
(also maxRetry is just minRetry*2, or minRetry<<1)

Reply:
nice catch guys 😍